Sources/Workspace/WorkspaceConfiguration.swift

```diff
@@ -50,39 +47,283 @@ extension Workspace {
+        ///
+        /// - Parameters:
+        ///   - localMirrorFile: Path to the workspace mirrors configuration file
+        ///   - sharedMirrorFile: Path to the shared mirrors configuration file, defaults to the standard location.
+        ///   - fileSystem: The file system to use.
+        public init(
+            localMirrorFile: AbsolutePath,
+            sharedMirrorFile: AbsolutePath?,
+            fileSystem: FileSystem
+        ) throws {
+            self.localMirrors = .init(path: localMirrorFile, fileSystem: fileSystem, deleteWhenEmpty: true)
+            self.sharedMirrors = sharedMirrorFile.map { .init(path: $0, fileSystem: fileSystem, deleteWhenEmpty: false) }
+            self.fileSystem = fileSystem
+            // computes the initial mirrors
+            self._mirrors = DependencyMirrors()
+            try self.computeMirrors()
+        }
+
+        @discardableResult
+        public func applyLocal(handler: (inout DependencyMirrors) throws -> Void) throws -> DependencyMirrors {
+            try self.localMirrors.apply(handler: handler)
+            try self.computeMirrors()
+            return self.mirrors
+        }
+
+        @discardableResult
+        public func applyShared(handler: (inout DependencyMirrors) throws -> Void) throws -> DependencyMirrors {
+            guard let sharedMirrors = self.sharedMirrors else {
+                throw InternalError("shared mirrors not configured")
+            }
+            try sharedMirrors.apply(handler: handler)
+            try self.computeMirrors()
+            return self.mirrors
+        }
+
+        // 👀 is this correct, or should we merge?
+        // mutating the state we hold since we are passing it by reference to the workspace
+        // access should be done using a lock
+        private func computeMirrors() throws {
+            try self.lock.withLock {
+                self._mirrors.removeAll()
+
+                // prefer local mirrors to shared ones
+                let local = try self.localMirrors.mirrors()
+                if !local.isEmpty {
+                    self._mirrors.append(contentsOf: local)
+                    return
+                }
+
+                // use shared if local was not found or empty
+                let shared = try self.sharedMirrors?.mirrors() ?? [:]
+                if !shared.isEmpty {
+                    self._mirrors.append(contentsOf: shared)
+                }
+            }
+        }
+    }
+
+    public struct Mirrors {
+        private let path: AbsolutePath
+        private let fileSystem: FileSystem
+        private let deleteWhenEmpty: Bool
+
+        public init(path: AbsolutePath, fileSystem: FileSystem, deleteWhenEmpty: Bool) {
+            self.path = path
+            self.fileSystem = fileSystem
+            self.deleteWhenEmpty = deleteWhenEmpty
+        }
+
+        /// The mirrors in this configuration
+        public func mirrors() throws -> DependencyMirrors {
```

Comment:
What do you think about renaming this to `load()` and moving the existing `load` helper method to an extension on `DependencyMirrors`? (It seems to me that the logic of initializing `mirrorsMap` on [line 285](https://github.com/apple/swift-package-manager/pull/3670/files#diff-8ea56cbcd3f938d28b68dee635b50b8afb4616cbc002044be94cf268743b7323R285) should be a concern of `DependencyMirrors`)

Reply:
turning `DependencyMirrors` into a pure data model is intentional, such that the caller can pass in mirrors that do not necessarily come from a configuration file. as such the load/save functionality moved into this utility which manages the mirrors configuration file. 

that said, I dont love the `mirrors()` func name either, since we use `apply` for the mutation, should we use `get()`?

Reply:
Sounds good. 👍 for `get()` as an alternative to `mirrors()`.

Reply:
I think serialization / coding wouldn't be _too_ onerous to the data model, so long as we kept the file management parts out of it. But that comes down to a personal preference, and it's really fine either way.

As far as naming, I think `get()` is fine for the current type name. If it changes to something like `MirrorsConfigurationFile` or anything emphasizing the "fileness" of it all, I'd stick with `mirrors()`.